### PR TITLE
#5 Created Common Controller

### LINKED ISSUE #5: CommonController: API to userProfile
The following API endpoints must be implemented in 'CommonController' class:

 **userProfile - "/userprofile/{userId}"**

This endpoint is used to get the details of any user in the Quora Application. This endpoint can be accessed by any user in the application.

- It should be a GET request

- This endpoint must request the path variable 'userId' as a string for the corresponding user profile to be retrieved and access token of the signed in user as a string in authorization Request Header.

- If the access token provided by the user does not exist in the database throw 'AuthorizationFailedException' with the message code - 'ATHR-001' and message - 'User has not signed in'.

- If the user has signed out, throw "AuthorizationFailedException" with the message code -'ATHR-002' and message -'User is signed out.Sign in first to get user details' .

- If the user with uuid whose profile is to be retrieved does not exist in the database, throw 'UserNotFoundException' with the message code -'USR-001' and message -'User with entered uuid does not exist'.

- Else, return all the details of the user from the database in the JSON response with the corresponding HTTP status.
- Read more @https://learn.upgrad.com/v/course/867/session/140118/segment/759581

### PR DESCRIPTION
**Pull request by soniyakocher**
Implemented getUser API endpoint which give user detailed response . All Common controller test cases are passed successfully

### CODE REVIEW

File: quora-service/src/main/java/com/upgrad/quora/service/business/CommonService.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+public class CommonService {
+
+    @Autowired
+    private UserDao userDao;
+
+    /**
+     *
```
Comment (cushivas) [inline quora-service/src/main/java/com/upgrad/quora/service/business/CommonService.java:20]: Please add description about this method.

**Review by cushivas — CHANGES_REQUESTED** [review]
Please add description

File: quora-api/src/main/java/com/upgrad/quora/api/controller/CommonController.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+     *
+     * @param authorizationToken
+     * @param userUuid
+     * @return UserDetailResponse with Http status
+     * @throws UserNotFoundException        if the authorization details of user are not found
+     * @throws AuthorizationFailedException if user is not authorised to get details of user
+     */
+
+    @RequestMapping(method = RequestMethod.GET, path = "/userprofile/{userId}", produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
+    public ResponseEntity<UserDetailsResponse> getUser(@PathVariable("userId") final String userUuid, @RequestHeader("authorization") final String authorizationToken)
```
Comment (cushivas) [inline quora-api/src/main/java/com/upgrad/quora/api/controller/CommonController.java:33]: Rename userUuid to userId

File: quora-service/src/main/java/com/upgrad/quora/service/business/CommonService.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+            throw new AuthorizationFailedException("ATHR-001", "User has not signed in");
+        }
+
+        //Check if user has signOut
+        if (userAuthTokenEntity.getLogoutAt() != null && userAuthTokenEntity.getLogoutAt().isAfter(userAuthTokenEntity.getLoginAt())) {
+            throw new AuthorizationFailedException("ATHR-002", "User is signed out.Sign in first to get user details");
+        }
+
+        //Check if uuid exist in database
+        UserEntity userEntityByUuid = userDao.getUserByUuid(userUuid);
```
Comment (cushivas) [inline quora-service/src/main/java/com/upgrad/quora/service/business/CommonService.java:44]: you need to pick uuid from userTokenEntity.

**Review by cushivas — CHANGES_REQUESTED** [review]
More comments inside

File: quora-service/src/main/java/com/upgrad/quora/service/business/CommonService.java (hunk 1 above)
Comment (soniyakocher) [reply quora-service/src/main/java/com/upgrad/quora/service/business/CommonService.java:20]: Added

File: quora-api/src/main/java/com/upgrad/quora/api/controller/CommonController.java (hunk 2 above)
Comment (soniyakocher) [reply quora-api/src/main/java/com/upgrad/quora/api/controller/CommonController.java:33]: it should be userUuid

File: quora-service/src/main/java/com/upgrad/quora/service/business/CommonService.java (hunk 3 above)
Comment (soniyakocher) [reply quora-service/src/main/java/com/upgrad/quora/service/business/CommonService.java:44]: we need to pick uuid from userEntity

**Review by cushivas — APPROVED** [review]
Looks good..

**Pull request merged**